bundle/edu.gemini.lchquery.servlet/src/main/scala/edu/gemini/lchquery/servlet/LchQueryParam.scala

```diff
@@ -0,0 +1,232 @@
+import edu.gemini.spModel.`type`.DisplayableSpType
+import edu.gemini.spModel.ao.{AOConstants, AOTreeUtil}
+import edu.gemini.spModel.data.YesNoType
+import edu.gemini.spModel.gemini.altair.{AltairParams, InstAltair}
+import edu.gemini.spModel.gemini.obscomp.{SPProgram, SPSiteQuality}
+import edu.gemini.spModel.gemini.obscomp.SPProgram.Active
+import edu.gemini.spModel.obs.{ObsClassService, ObservationStatus, SPObservation}
+import edu.gemini.spModel.obsclass.ObsClass
+import edu.gemini.spModel.obscomp.SPInstObsComp
+import edu.gemini.spModel.target.env.TargetEnvironment
+import edu.gemini.spModel.target.obsComp.TargetObsComp
+import edu.gemini.spModel.too.Too
+import edu.gemini.spModel.util.SPTreeUtil
+
+import scala.collection.JavaConverters._
+import scala.util.matching.Regex
+import scalaz._
+import Scalaz._
+
+
+sealed trait ValueMatcher[A] {
+  def matches(expression: String, x: A): Boolean
+}
+
+object ValueMatcher {
+  private[servlet] implicit class ToRegex(val expression: String) extends AnyVal {
+    def toRegex: Regex = ("^(?i)" +
+      (expression.contains("|") ? s"($expression)" | expression).
+        replaceAllLiterally("+", "\\+").
+        replaceAllLiterally("*", ".*").
+        replaceAllLiterally("%", ".*").
+        replaceAllLiterally("?", ".")
+      + "$").r
+  }
+
+  abstract class StringValueMatcher[A] extends ValueMatcher[A] {
+    protected def extractor(a: A): Option[String]
+
+    override def matches(expression: String, x: A): Boolean = (for {
+      r <- Option(expression).map(_.toRegex)
+      m <- Option(x).flatMap(extractor)
+    } yield r.findFirstMatchIn(m).isDefined).getOrElse(false)
+  }
+
+
+  abstract class BooleanValueMatcher[A, B <: DisplayableSpType] extends ValueMatcher[A] {
+    protected def extractor(a: A): Option[B]
+
+    private def transform(expression: String): String = expression.toLowerCase match {
+      case "true" => YesNoType.YES.displayValue()
+      case "false" => YesNoType.NO.displayValue()
+      case _ => expression
+    }
+
+    override def matches(expression: String, x: A): Boolean = (for {
+      r <- Option(expression).map(transform).map(_.toRegex)
+      m <- Option(x).flatMap(extractor).map(_.displayValue)
+    } yield r.findFirstMatchIn(m).isDefined).getOrElse(false)
+  }
+}
+
+
+sealed case class LchQueryParam[A](name: String, v: ValueMatcher[A])
+
+object LchQueryParam {
+  import ValueMatcher._
+
+  private[servlet] implicit class ToYesNo(val b: Boolean) extends AnyVal {
+    def toYesNo: YesNoType =
+      b ? YesNoType.YES | YesNoType.NO
+  }
+
+
+  private[servlet] implicit class ISPProgramExtractors(val prog: ISPProgram) extends AnyVal {
+    def toSPProg: Option[SPProgram] =
+      Option(prog).map(_.getDataObject.asInstanceOf[SPProgram])
+
+    def semester: Option[String] = {
+      val semesterRe = "(?i)\\-(2\\d{3}[AB])".r
+      Option(prog.getProgramID).flatMap(id => semesterRe.findFirstMatchIn(id.stringValue()).map(_.group(1)))
```

Comment:
Let's take advantage of `edu.gemini.spModel.core.ProgramId.parse(s: String)` so that you don't have to define another regular expression.